src/main/java/com/lineate/testyourlexicon/services/GameService.java

```diff
@@ -47,13 +54,116 @@ public SupportedLanguagesDto supportedLanguages() {
+    game.setStepsLeft(game.getStepsLeft() - 1);
+    game.setCurrentQuestionId(questionEntity.getId());
+    return gameRepository.save(game);
+  }
+
+  public StepDto userActiveGameNextStep(User user, long gameId) {
+    Game game = validateGameForUser(user, gameId);
+    Question question = generateRandomQuestion(user.getGameConfiguration(), game);
+    QuestionEntity questionEntity = saveQuestionInDatabase(question, game);
+    game = updateGameCurrentQuestion(game, questionEntity);
+    log.info("Generated step {'user': {}, 'game_id': {}, 'question_id': {}}",
+        user.getId(), game.getGameId(), questionEntity.getId());
+    return new StepDto(question, gameId);
+  }
+
+  public QuestionEntity getQuestion(Long questionId) {
+    if (questionId == null) {
+      throw new GeneralMessageException("Couldn't not find active question");
+    }
+    Optional<QuestionEntity> questionEntity =
+      questionRepository.findById(questionId);
+    if (questionEntity.isEmpty()) {
+      throw new GeneralMessageException("Invalid question id");
+    }
+    return questionEntity.get();
+  }
+
+  public AnswerResponseDto userActiveGameAnswer(User user,
+                                   AnswerRequestDto answerRequestDto,
+                                   long gameId) {
+    Game game = validateGameForUser(user, gameId);
+    QuestionEntity questionEntity = getQuestion(game.getCurrentQuestionId());
+
+    // Get correct answer
+    Long translationId = questionEntity.getTranslationId();
+    String language = user.getGameConfiguration().getTranslateTo();
+    String correctAnswer =
+         translationRepository.languageDefinitionGivenIdAndLanguage(translationId, language);
+
+    // Check if users answer is the same as correct one
+    boolean guessed = correctAnswer.equalsIgnoreCase(answerRequestDto.getAnswer());
+    if (guessed) {
+      questionEntity.setGuessed(true);
+      questionRepository.save(questionEntity);
+    }
+    game.setCurrentQuestionId(null);
+    gameRepository.save(game);
+    log.info("User answered a step {'user': {}, 'game_id': {}, 'question': {}}",
+        user.getId(), game.getGameId(), questionEntity.getId());
+    return AnswerResponseDto.builder()
+      .guessed(guessed)
+      .correctAnswer(correctAnswer)
+      .userAnswer(answerRequestDto.getAnswer())
+      .build();
+  }
+
+  private void checkIfGameActiveForUser(User user) {
+    if (gameRepository.getUserActiveGame(user).isPresent()) {
+      throw new GeneralMessageException("A game is already started for the user");
+    }
+  }
+
+  private Game validateGameForUser(User user, Long gameId) {
+    Optional<Game> gameOptional = gameRepository.findById(gameId);
+    if (gameOptional.isEmpty()) {
+      log.info("User tried to access game with invalid id "
+          + "{'user': {}, 'game_id': {}}", user.getId(), gameId);
+      throw new GeneralMessageException("Game id not valid");
+    }
```

Comment:
I think you can use Optional's `orElseThrow` and get rid of this `if`. Also it will allow us to get rid of 157th line